drivers/gpu/drm/panel/panel-simple.c

```diff
@@ -4681,23 +4684,123 @@ static const struct of_device_id dsi_of_match[] = {
+		return -ENOMEM;
+
+	timing = devm_kzalloc(dev, sizeof(*timing), GFP_KERNEL);
+	if (!timing)
+		return -ENOMEM;
+
+	ret = of_get_display_timing(np, "panel-timing", timing);
+	if (ret < 0) {
+		dev_err(dev, "%pOF: no panel-timing node found for \"panel-dsi\" binding\n",
+			np);
+		return ret;
+	}
+
+	desc->timings = timing;
+	desc->num_timings = 1;
+
+	of_property_read_u32(np, "width-mm", &desc->size.width);
+	of_property_read_u32(np, "height-mm", &desc->size.height);
+
+	of_property_read_u32(np, "lanes", &desc_dsi->lanes);
+
+	of_property_read_string(np, "dsi-color-format", &dsi_color_format);
+	if (!strcmp(dsi_color_format, "RGB888")) {
+		desc_dsi->format = MIPI_DSI_FMT_RGB888;
+		desc->bpc = 8;
+	} else if (!strcmp(dsi_color_format, "RGB565")) {
+		desc_dsi->format = MIPI_DSI_FMT_RGB565;
+		desc->bpc = 6;
+	} else if (!strcmp(dsi_color_format, "RGB666")) {
+		desc_dsi->format = MIPI_DSI_FMT_RGB666;
+		desc->bpc = 6;
+	} else if (!strcmp(dsi_color_format, "RGB666_PACKED")) {
+		desc_dsi->format = MIPI_DSI_FMT_RGB666_PACKED;
+		desc->bpc = 6;
+	}
+
+	of_property_read_string(np, "mode", &dsi_mode_flags);
+	if (!strcmp(dsi_mode_flags, "MODE_VIDEO"))
+		desc_dsi->flags = MIPI_DSI_MODE_VIDEO;
+	else if (!strcmp(dsi_mode_flags, "MODE_VIDEO_BURST"))
+		desc_dsi->flags = MIPI_DSI_MODE_VIDEO_BURST;
+	else if (!strcmp(dsi_mode_flags, "MODE_VIDEO_SYNC_PULSE"))
+		desc_dsi->flags = MIPI_DSI_MODE_VIDEO_SYNC_PULSE;
+	else if (!strcmp(dsi_mode_flags, "MODE_VIDEO_AUTO_VERT"))
+		desc_dsi->flags = MIPI_DSI_MODE_VIDEO_AUTO_VERT;
+	else if (!strcmp(dsi_mode_flags, "MODE_VIDEO_HSE"))
+		desc_dsi->flags = MIPI_DSI_MODE_VIDEO_HSE;
+	else if (!strcmp(dsi_mode_flags, "MODE_VIDEO_NO_HFP"))
+		desc_dsi->flags = MIPI_DSI_MODE_VIDEO_NO_HFP;
+	else if (!strcmp(dsi_mode_flags, "MODE_VIDEO_NO_HBP"))
+		desc_dsi->flags = MIPI_DSI_MODE_VIDEO_NO_HBP;
+	else if (!strcmp(dsi_mode_flags, "MODE_VIDEO_NO_HSA"))
+		desc_dsi->flags = MIPI_DSI_MODE_VIDEO_NO_HSA;
+	else if (!strcmp(dsi_mode_flags, "MODE_VSYNC_FLUSH"))
+		desc_dsi->flags = MIPI_DSI_MODE_VSYNC_FLUSH;
+	else if (!strcmp(dsi_mode_flags, "MODE_NO_EOT_PACKET"))
+		desc_dsi->flags = MIPI_DSI_MODE_NO_EOT_PACKET;
+	else if (!strcmp(dsi_mode_flags, "CLOCK_NON_CONTINUOUS"))
+		desc_dsi->flags = MIPI_DSI_CLOCK_NON_CONTINUOUS;
+	else if (!strcmp(dsi_mode_flags, "MODE_LPM"))
+		desc_dsi->flags = MIPI_DSI_MODE_LPM;
+	else if (!strcmp(dsi_mode_flags, "HS_PKT_END_ALIGNED"))
+		desc_dsi->flags = MIPI_DSI_HS_PKT_END_ALIGNED;
+
+
+	desc->connector_type = DRM_MODE_CONNECTOR_DSI;
+	desc_dsi->desc = *desc;
+
+	return 0;
+}
+
 static int panel_simple_dsi_probe(struct mipi_dsi_device *dsi)
 {
 	const struct panel_desc_dsi *desc;
+	struct panel_desc_dsi *dt_desc;
 	const struct of_device_id *id;
 	int err;
 
 	id = of_match_node(dsi_of_match, dsi->dev.of_node);
 	if (!id)
 		return -ENODEV;
 
-	desc = id->data;
+	if (id->data == &panel_dsi) {
+		/* Handle the generic panel-dsi binding */
+		err = panel_dsi_dt_probe(&dsi->dev, dt_desc);
```

Comment:
The handlng of dt_desc here looks suspicious.

Defined as an uninitialised pointer, that uninitialised value is passed to panel_dsi_dt_probe which dereferences it and writes the desc field. We then store that uninitialised value in desc for use later :-/

Reply:
I might be missing what you mean but panel_dsi_dt_probe allocates the desc struct that gets assigned?

Reply:
Oh wait yea no I think you have a point, the outer member pointer itself does not get allocated. This might be running on undefined behaviour right now, otherwise it should blow up which it weirdly doesn't atm.

What would you think about making dt_desc static? My main gripe there was that the panel definitions are supposed to be treated as const there so I can't reuse panel_dsi which would be the better way really.

Reply:
You can't reuse `panel_dsi` as you could have 2 displays configured through the same `panel-dsi` compatible.

You are right that there is a conflict that `panel_simple_probe` gets a dummy desc as panel_dpi, and then `panel_dpi_probe` handles allocating the desc. That bit almost does want to be pulled out in order to make it potentially common.

Your current implementation does all the work before `panel_simple_probe` gets a look in.

Reply:
Oh right, I guess that might need fixing then too for panel-dpi, that is currently re-useing the single dummy struct :)
Though I'm not sure how you would handle multiple panels like this properly that does not rely on hard coded dummy panels to refer to. I think that would require some deeper changes to remove the assumption that a panel is in some way already statically defined. I might need to test this if it might already work unintentionally, the whole driver is a bit of a rabbit hole of structs being passed around and torn apart.

Reply:
panel_dpi is fine as `desc` gets swapped for `panel->desc`, which is allocated in `panel_dpi_probe`.

Now that you devm_kzalloc the desc, I think you're OK too.

Reply:
I would like to see the duplication removed, but only if you're committing to upstream the differences.
Handling major changes to files such as this that get a fair amount of churn is not easy.

Reply:
As in for an upstream submission you would like to see the larger change that abstracts the device tree side timing definition?